### PR TITLE
Quic Develop

### PR DESCRIPTION
**Pull request by wanghaEMQ**

### CODE REVIEW

**Comment by JaylinYu** [comment]
What a huge PR

**Review by JaylinYu — APPROVED** [review]
Not finished yet...  but I trust u

File: src/mqtt/transport/quic/mqtt_quic.c [hunk 1]
```diff
@@ -0,0 +1,1869 @@
...
+	nni_mtx_init(&ep->mtx);
+	NNI_LIST_INIT(&ep->busypipes, mqtt_quictran_pipe, node);
+	NNI_LIST_INIT(&ep->waitpipes, mqtt_quictran_pipe, node);
+	NNI_LIST_INIT(&ep->negopipes, mqtt_quictran_pipe, node);
+
+	ep->proto       = nni_sock_proto_id(sock);
+	ep->url         = url;
+	ep->connmsg     = NULL;
+	ep->reason_code = 0;
+	ep->property    = NULL;
```
Comment (JaylinYu) [inline src/mqtt/transport/quic/mqtt_quic.c:1502]: 0rtt should belong to ep?

File: src/mqtt/transport/quic/mqtt_quic.c [hunk 2]
```diff
@@ -0,0 +1,1869 @@
...
+	int              rv;
+
+	if (nni_aio_begin(aio) != 0) {
+		return;
+	}
+	if (ep->backoff != 0) {
+		ep->backoff = ep->backoff * 2;
+		ep->backoff = ep->backoff > ep->backoff_max
+		    ? nni_random() % 2000
+		    : ep->backoff;
```
Comment (JaylinYu) [inline src/mqtt/transport/quic/mqtt_quic.c:1595]: I see . it is Cool

File: src/mqtt/transport/quic/mqtt_quic.c [hunk 3]
```diff
@@ -0,0 +1,1869 @@
...
+	rv = nni_copyin_ptr(&ep->connmsg, v, sz, t);
+	nni_mtx_unlock(&ep->mtx);
+
+	return (rv);
+}
+
+// NanoSDK use exponential backoff strategy as default
+// Backoff for random time that exponentially curving
+static int
+mqtt_quictran_ep_set_reconnect_backoff(void *arg, const void *v, size_t sz, nni_opt_type t)
```
Comment (JaylinYu) [inline src/mqtt/transport/quic/mqtt_quic.c:1694]: do the same for 0rtt

File: src/mqtt/transport/quic/mqtt_quic.c [hunk 4]
```diff
@@ -0,0 +1,1869 @@
...
+	rv = nni_stream_dialer_get(ep->dialer, name, buf, szp, t);
+	if (rv == NNG_ENOTSUP) {
+		rv = nni_getopt(mqtt_quictran_ep_opts, name, ep, buf, szp, t);
+	}
+	return (rv);
+}
+
+static int
+mqtt_quictran_dialer_setopt(
+    void *arg, const char *name, const void *buf, size_t sz, nni_type t)
```
Comment (JaylinYu) [inline src/mqtt/transport/quic/mqtt_quic.c:1813]: does this really working now?

File: src/supplemental/mqtt/mqtt_codec.c
Comment (JaylinYu) [inline src/supplemental/mqtt/mqtt_codec.c:1]: discard this change plz

File: src/supplemental/quic/msquic_dial.c [hunk 5]
```diff
@@ -0,0 +1,1256 @@
...
+	uint8_t         reason_code;
+
+	nni_reap_node   reap;
+};
+
+static const QUIC_API_TABLE *MsQuic = NULL;
+
+// Config for msquic
+static const QUIC_REGISTRATION_CONFIG quic_reg_config = {
+	"mqtt",
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:85]: how about exposing msquic configurations here?

File: src/supplemental/quic/msquic_dial.c [hunk 6]
```diff
@@ -0,0 +1,1256 @@
...
+	bool             ismain = false;
+	int              rv;
+
+	if (nni_aio_begin(aio) != 0) {
+		return;
+	}
+
+	nni_atomic_inc64(&d->ref);
+
+	// Create a connection whenever dial. So it's okey. right?
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:263]: incorrect if you suggest we use dialer open to start a new stream

File: src/supplemental/quic/msquic_dial.c [hunk 7]
```diff
@@ -0,0 +1,1256 @@
...
+	}
+
+	// pass c to the dialer for getting it in msquic_strm_open
+	d->currcon = c;
+	// set c to provdata. Which is helpful for quic dialer close.
+	nni_aio_set_prov_data(aio, c);
+	c->dial_aio = aio;
+
+	if (ismain) {
+		// Make a quic connection to the url.
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:301]: cool

File: src/supplemental/quic/msquic_dial.c [hunk 8]
```diff
@@ -0,0 +1,1256 @@
...
+void
+nni_quic_dialer_close(void *arg)
+{
+	nni_quic_dialer *d = arg;
+
+	nni_mtx_lock(&d->mtx);
+	if (!d->closed) {
+		nni_aio *aio;
+		d->closed = true;
+		while ((aio = nni_list_first(&d->connq)) != NULL) {
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:343]: attentions here

File: src/supplemental/quic/msquic_dial.c [hunk 9]
```diff
@@ -0,0 +1,1256 @@
...
+	for (int i=0; i<naiov; ++i) {
+		log_debug("buf%d sz %d", i, aiov[i].iov_len);
+		buf[i].Buffer = aiov[i].iov_buf;
+		buf[i].Length = aiov[i].iov_len;
+		n += aiov[i].iov_len;
+	}
+	nni_aio_set_input(aio, 0, buf);
+
+	if (QUIC_FAILED(rv = MsQuic->StreamSend(c->qstrm, buf,
+	                naiov, QUIC_SEND_FLAG_NONE, aio))) {
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:569]: no corresponding action of this aio

File: src/supplemental/quic/msquic_dial.c [hunk 10]
```diff
@@ -0,0 +1,1256 @@
...
+			    stream, Event->SEND_COMPLETE.Canceled);
+		}
+		// Priority msg send
+		if ((aio = Event->SEND_COMPLETE.ClientContext) != NULL) {
+			Event->SEND_COMPLETE.ClientContext = NULL;
+			nni_aio_finish(aio, 0, nni_aio_count(aio));
+			break;
+		}
+		// Ordinary send
+		quic_cb(QUIC_STREAM_EVENT_SEND_COMPLETE, c);
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:883]: not sure why do you use embedded cb here, plz make it inline

File: src/supplemental/quic/msquic_dial.c [hunk 11]
```diff
@@ -0,0 +1,1256 @@
...
+msquic_strm_fini(HQUIC qstrm)
+{
+	MsQuic->StreamClose(qstrm);
+}
+
+static void
+msquic_strm_recv_start(HQUIC qstrm)
+{
+	MsQuic->StreamReceiveSetEnabled(qstrm, TRUE);
+}
```
Comment (JaylinYu) [inline src/supplemental/quic/msquic_dial.c:1256]: pretty cool

File: src/supplemental/quic/quic_api.c [hunk 12]
```diff
@@ -27,6 +28,610 @@
...
+	nni_mtx_lock(&d->mtx);
+	d->qmax_ack_delay_ms = n;
+	d->settings.IsSet.MaxAckDelayMs = TRUE;
+	d->settings.MaxAckDelayMs = n;
+	nni_mtx_unlock(&d->mtx);
+	return 0;
+}
+
+static int
+quic_dialer_set_initial_rtt_ms(void *arg, void *buf, size_t sz, nni_type t)
```
Comment (JaylinYu) [inline src/supplemental/quic/quic_api.c:279]: Oh I see.
We can simply this

File: src/supplemental/quic/quic_api.c [hunk 13]
```diff
@@ -27,6 +28,610 @@
...
+		return rv;
+	}
+
+	nni_mtx_lock(&d->mtx);
+	d->enable_0rtt = b;
+	nni_mtx_unlock(&d->mtx);
+	return 0;
+}
+
+static const nni_option quic_dialer_options[] = {
```
Comment (JaylinYu) [inline src/supplemental/quic/quic_api.c:452]: cool

File: src/tools/perf/CMakeLists.txt [hunk 14]
```diff
@@ -14,7 +14,12 @@ nng_directory(perf)
 if (NNG_TESTS)
      macro (add_nng_perf NAME)
         add_executable (${NAME} perf.c)
-        target_link_libraries (${NAME} nng nng_private)
+        if(NNG_ENABLE_QUIC)
+            target_link_libraries (${NAME} nng nng_private msquic OpenSSLQuic)
```
Comment (JaylinYu) [inline src/tools/perf/CMakeLists.txt:18]: working now?

**Review by JaylinYu — APPROVED** [review]
still pretty cool, cant help having a try

**Comment by JaylinYu** [comment]
which branch in nanomq should I using for this?

File: include/nng/nng.h [hunk 15]
```diff
@@ -727,6 +727,25 @@ NNG_DECL nng_listener nng_pipe_listener(nng_pipe);
...
 
+// NNG-QUIC
+#define NNG_OPT_QUIC_ENABLE_0RTT "quic-0rtt"
+#define NNG_OPT_QUIC_ENABLE_MULTISTREAM "quic-multistream"
+#define NNG_OPT_QUIC_IDLE_TIMEOUT "quic-idle-timeout"
+#define NNG_OPT_QUIC_KEEPALIVE "quic-keepalive"
+#define NNG_OPT_QUIC_CONNECT_TIMEOUT "quic-connect-timeout"
+#define NNG_OPT_QUIC_DISCONNECT_TIMEOUT "quic-disconnect-timeout"
+#define NNG_OPT_QUIC_SEND_IDLE_TIMEOUT "quic-send-idle-timeout"
+#define NNG_OPT_QUIC_INITIAL_RTT_MS "quic-initial-ms"
```
[the remark this answers is not in the document]
Comment (wanghaEMQ) [reply include/nng/nng.h:738]: Yes

File: include/nng/mqtt/mqtt_quic.h [hunk 16]
```diff
@@ -20,7 +20,8 @@ extern "C" {
 #if defined(SUPP_QUIC)
 
 NNG_DECL int nng_mqttv5_quic_client_open(nng_socket *, const char *url);
-NNG_DECL int nng_mqtt_quic_client_open(nng_socket *, const char *url);
+NNG_DECL int nng_mqtt_quic_client_open(nng_socket *);
```
[the remark this answers is not in the document]
Comment (wanghaEMQ) [reply include/nng/mqtt/mqtt_quic.h:23]: Yes. There is a similar interface in tcp

File: src/mqtt/transport/quic/mqtt_quic.c [hunk 17]
```diff
@@ -0,0 +1,1869 @@
+//
```
[the remark this answers is not in the document]
Comment (wanghaEMQ) [reply src/mqtt/transport/quic/mqtt_quic.c:1]: Disagree. The transport layer and protocol layer are closely related in the protocol self, test and interoperation

File: src/supplemental/quic/msquic_dial.c (hunk 10 above)
Comment (wanghaEMQ) [reply src/supplemental/quic/msquic_dial.c:883]: Split msquic and nng

Comment (wanghaEMQ) [reply src/supplemental/quic/msquic_dial.c:883]: And easy to read

**Comment by wanghaEMQ** [comment]
> which branch in nanomq should I using for this?

https://github.com/emqx/nanomq/pull/1415

**Review by JaylinYu — APPROVED** [review]

File: src/core/stream.c [hunk 18]
```diff
@@ -209,6 +210,23 @@ static struct {
 	    .dialer_alloc   = nni_ws_dialer_alloc,
 	    .listener_alloc = nni_ws_listener_alloc,
 	},
+// #ifdef NNG_HAVE_QUIC_SOCKETS
+	{
+	    .scheme         = "quic",
+	    .dialer_alloc   = nni_quic_dialer_alloc,
```
[the remark this answers is not in the document]
Comment (wanghaEMQ) [reply src/core/stream.c:219]: No. I'm just being lazy here hhh

**Pull request merged**